homeassistant/components/refoss/switch.py

```diff
@@ -0,0 +1,101 @@
+"""Switch for Refoss."""
+
+from __future__ import annotations
+
+from typing import Any
+
+from refoss_ha.controller.device import BaseDevice
+from refoss_ha.controller.toggle import ToggleXMix
+
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import HomeAssistantRefossData
+from .const import DOMAIN, LOGGER, REFOSS_DISCOVERY_NEW
+from .device import RefossEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """async_setup_entry."""
+    hass_data: HomeAssistantRefossData = hass.data[DOMAIN][entry.entry_id]
+
+    @callback
+    def entity_add_callback(device_ids: list[str]) -> None:
+        """entity_add_callback."""
+        new_entities = []
+        try:
+            for uuid in device_ids:
+                device = hass_data.device_manager.base_device_map[uuid]
+                if device is None:
+                    continue
+                if not isinstance(device, ToggleXMix):
+                    continue
+
+                if len(device.channels) == 0:
+                    continue
+
+                for channel in device.channels:
+                    w = SwitchEntityWrapper(device=device, channel=channel)
+                    new_entities.append(w)
+
+        except Exception as e:
+            LOGGER.debug(f"setup switch fail,err:{e}")
+            raise e
+
+        async_add_entities(new_entities, True)
+
+    entity_add_callback([*hass_data.device_manager.base_device_map])
+
+    entry.async_on_unload(
+        async_dispatcher_connect(hass, REFOSS_DISCOVERY_NEW, entity_add_callback)
+    )
```

Comment:
Since this isn't a hub integration, we want one device per config entry instead of having multiple devices come in the same config entry.

This allows HA to cache the ip addresses and update them based on unique id when they change.

Check how `tplink` and `lifx` are written